mis/pipeline/pulse.py

```diff
@@ -20,59 +26,119 @@ class BasePulseShaper(ABC):
-        """
-        Method to return a simple constant waveform pulse
-        """
-        wf = ConstantWaveform(duration=1000, value=1.0)
-        pulser_pulse = PulserPulse.ConstantDetuning(amplitude=wf, detuning=0.0, phase=0.0)
-
-        self.pulse = Pulse(pulse=pulser_pulse)
-        return self.pulse
+@dataclass
+class _Bounds:
+    maximum_amplitude: float
+    final_detuning: float
 
 
-def get_pulse_shaper(instance: Any, config: SolverConfig) -> BasePulseShaper:
+class DefaultPulseShaper(BasePulseShaper):
+    """
+    A simple pulse shaper.
     """
-    Method that returns the correct PulseShaper based on configuration.
-    The correct pulse shaping method can be identified using the config, and an
-    object of this pulseshaper can be returned using this function.
 
-    Args:
-        instance (MISInstance): The MIS problem to embed.
-        config (Device): The quantum device to target.
+    def _get_interactions(
+        self, pos: np.ndarray, graph: nx.Graph, device: Device
+    ) -> tuple[list[float], list[float]]:
+        """Calculate the interaction strengths for connected and disconnected nodes.
 
-    Returns:
-        (BasePulseShaper): The representative Pulse Shaper object.
-    """
+        Args:
+            pos (np.ndarray): The position of the nodes.
+            graph (nx.Graph): The associated graph.
+            device (BaseDevice): Device used to calculate interaction coeff.
+
+        Returns:
+            tuple[list[float], list[float]]: Connected interactions, Disconnected interactions
+        """
 
-    return FirstPulseShaper(instance, config)
+        def calculate_edge_interaction(edge: tuple[int, int]) -> float:
+            pos_a, pos_b = pos[edge[0]], pos[edge[1]]
+            return float(device.interaction_coeff / (euclidean(pos_a, pos_b) ** 6))
+
+        connected = [calculate_edge_interaction(edge) for edge in graph.edges()]
+        disconnected = [calculate_edge_interaction(edge) for edge in nx.complement(graph).edges()]
+
+        return connected, disconnected
+
+    def _calc_bounds(self, reg: Register, graph: nx.Graph, device: Device) -> _Bounds:
+        _, disconnected = self._get_interactions(reg.register.sorted_coords, graph, device)
+        u_min, u_max = self._interaction_bounds(reg.register.sorted_coords, graph, device)
+        max_amp_device = device.channels["rydberg_global"].max_amp or np.inf
+        maximum_amplitude = min(max_amp_device, u_max + 0.8 * (u_min - u_max))
+
+        degree = graph.degree
+        assert isinstance(degree, DegreeView)
+        d_min = min(dict(degree).values())
+        d_max = max(dict(degree).values())
+        det_max_theory = (d_min / (d_min + 1)) * u_min
+        det_min_theory = sum(sorted(disconnected)[-d_max:])
+        det_final_theory = max([det_max_theory, det_min_theory])
+        det_max_device = device.channels["rydberg_global"].max_abs_detuning or np.inf
+        final_detuning = min(det_final_theory, det_max_device)
```

Comment:
Hm, I kinda get it but it might deserve a bit of more explanations.

Reply:
Yeah, so far, I've basically copied this from the notebook.

Reply:
Filed as #20 .